### DISCUSSION TITLE
Support userland data in SSR contexts

### DISCUSSION THREAD

**Discussion by cdaringe** [vercel/next.js#14396] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

To pass custom data into `getServerSideProps`, I have to mutate `req: IncomingRequest` in my custom-server.

## Describe the solution you'd like

For custom server users, there is business logic required before next.js begins its render cycle. If I need access to my own custom server resources inside of, say, `getServerSideProps` or `getInitialProps`, there is no way to thread my own data from the customer server into next. I have to mutate the raw `req` from node to pass data from my custom server and read that new field from within next.

Writing to memory that I do not own is a bad practice.

Consider adding a user-writable context, that allows me to pass data from my custom server into `next`. Much like `hapi` has `server.app` (`ApplicationContext`) or koa has `BaseContext` which users can augment, so too would I like to
augment the next.js context. In this respect, the context would look more like:

```ts
export type GetServerSidePropsContext<
  Q extends ParsedUrlQuery = ParsedUrlQuery,
  AppContext=void
> = {
  req: IncomingMessage
  res: ServerResponse
  ...snip
  app?: AppContext // 👀 a new field for me, by me!
}
```

To make that happen, there's quite a few touch points, but the lift could be
minimal, if all we did was app _optional arguments_. Here's an example flow
from the next handler to Context construction:

route ingress:

- https://github.com/vercel/next.js/blob/bef9b5610911532689dc251b7ceac72985e6b6dd/packages/next/next-server/server/next-server.ts#L245
- https://github.com/vercel/next.js/blob/bef9b5610911532689dc251b7ceac72985e6b6dd/packages/next/next-server/server/next-server.ts#L773
- https://github.com/vercel/next.js/blob/bef9b5610911532689dc251b7ceac72985e6b6dd/packages/next/next-server/server/router.ts#L159
- https://github.com/vercel/next.js/blob/bef9b5610911532689dc251b7ceac72985e6b6dd/packages/next/next-server/server/router.ts#L283
- https://github.com/vercel/next.js/blob/bef9b5610911532689dc251b7ceac72985e6b6dd/packages/next/next-server/server/router.ts#L19-L31

route setup & processing:
- https://github.com/vercel/next.js/blob/bef9b5610911532689dc251b7ceac72985e6b6dd/packages/next/server/next-dev-server.ts#L135
- eek! ran out of time. will try and pickup later :)

What the general sentiment towards such a feature? I'd be glad to send patches (including docs)
if the maintainers would be open to it, and if we could align around a strategy.

Thanks for the consideration and great tool.

## Describe alternatives you've considered

Augmenting `IncomingRequest` and mutating `req`

## Additional context

n/a

**Comment by timneutkens** [comment]
We are not planning to change the custom server api at this point in time, in general we recommend having all logic inside your Next.js application where possible.
